### PR TITLE
[WIP] Fixes #10: Operator "==" not working when the attributed class is in the same assembly as call site

### LINKED ISSUE #10: Operator "==" not working when the attributed class is in the same assembly as call site
Hi, it seems to me that the Fody generated operator "==" is not working when the attributed class with [Equals] is in the same assembly as call site.

I've done a runnable example which demonstrates the issue: https://github.com/IvanMacek/FodyEqualsFailExample

Call to the Equals method works as expected. 
When the class is implemented in the another assembly operator "==" works as expected.

Thank you for your time!

Cheers,
Ivan

### PR DESCRIPTION
**Pull request by BrunoJuchli**
This is a work-in-progress to fix #10 

This is a **breaking change**, and thus the major version was increased to 3.
I have documented the breaking change in Readme.md.

## Concept

To guarantee identical behavior for intra- and inter- assembly calls of the woven `==` and `!=` operators the source code *must* contain these operators. This change ensures that operators are only woven if this is the case. And it does so in a manner that can't be mistaken by the user.

### Default Case (Operators are woven)

For types with `[Equals]`, the operators are only woven when the user has implement stubs like:

    public static bool operator ==(Point left, Point right) => Operator.Weave();
    public static bool operator !=(Point left, Point right) => Operator.Weave();

In case he fails to do so, the weaver throws an exception.

### Opting Out
Opting out via `[Equals(DoNotAddEqualityOperators = true)]` has been preserved.
To prevent run-time errors due to stub implementations, the weaver checks that in case of opt out the operators do not (only) call `Operator.Weave()`.
If they do, the weaver throws an exception.

### Compiler Warnings
Whenever the implicit operator stubs  `==` and `!=` are added, compiler will generate warnings CS0660, CS0661, which tell you that you should also add custom `Equals` and `GetHashCode` implementations. For users, it's probably sensible to turn these warnings off when using Equals.Fody...?!
I've documented this in readme.md

## Merging
I suggest merging this with a squash-commit. Otherwise I can rewrite the PR history before merging back.

### CODE REVIEW

File: Equals.Fody/Operator.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        public static readonly Operator Equality = new Operator("op_Equality", true, "==");
+        public static readonly Operator Inequality = new Operator("op_Inequality", false, "!=");
+
+        private Operator(string methodName, bool isEquality, string sourceMethodName)
+        {
+            MethodName = methodName;
+            IsEquality = isEquality;
+            MethodSourceExample =
+                FormattableString.Invariant(
```
Comment (SimonCropp) [inline Equals.Fody/Operator.cs]: why is invariant needed here?

File: Equals.Fody/WeavingInstruction.cs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+using Fody;
+using Mono.Cecil;
+using Mono.Cecil.Cil;
+
+namespace Equals.Fody
+{
+    public class WeavingInstruction
+    {
+        private static readonly Instruction NotWeavingInstruction = Instruction.Create(OpCodes.Throw);
+        private readonly MethodDefinition _weaveMethod;
```
Comment (SimonCropp) [inline Equals.Fody/WeavingInstruction.cs]: kill the `private`, `readonly` and the _

File: Equals.Fody/WeavingInstruction.cs [hunk 3]
```diff
@@ -0,0 +1,83 @@
+﻿using System;
+using System.Linq;
+using Fody;
+using Mono.Cecil;
+using Mono.Cecil.Cil;
+
+namespace Equals.Fody
```
Comment (SimonCropp) [inline Equals.Fody/WeavingInstruction.cs]: namespace is redundant

File: Equals/Operator.cs [hunk 4]
```diff
@@ -0,0 +1,6 @@
+﻿using System;
+
+public static class Operator
+{
+    public static bool Weave() => throw new Exception("Equals.Fody was supposed to replace this method call with an implementation. Either weaving has not worked or you have called this method from an unsupported place. Only supported places are implementations of the `==` and `!=` operators.");
```
Comment (SimonCropp) [inline Equals/Operator.cs]: Equals.Fody was supposed to replace this method call with an implementation. Either weaving has not worked or the method has been invoked in an unsupported manner. Only `==` and `!=` operators are supported.

**Comment by SimonCropp** [comment]
@BrunoJuchli nice work. as a small thanks, can u invoice 30$ here https://opencollective.com/fody/expenses/new. or if that is too much hassle, nominate a charity of your choice and i will handle it in your name

**Comment by BrunoJuchli** [comment]
@SimonCropp 
Thanks a lot! Sorry for the formatting stuff.
Code was a contribution of [EROWA AG](https://opencollective.com/erowa-ag).
I can't take any money for that.
Pay it out to yourself for your continued support?!

**Pull request merged**